core/server/master/src/main/java/alluxio/master/file/meta/MountTable.java

```diff
@@ -105,10 +105,40 @@ public MountTable(UfsManager ufsManager, MountInfo rootMountInfo) {
    */
   public void add(Supplier<JournalContext> journalContext, AlluxioURI alluxioUri, AlluxioURI ufsUri,
       long mountId, MountPOptions options) throws FileAlreadyExistsException, InvalidPathException {
+    // validate the Mount operation first, error will be thrown if the operation is invalid
+    validateMountPoint(alluxioUri, ufsUri);
+
     String alluxioPath = alluxioUri.getPath().isEmpty() ? "/" : alluxioUri.getPath();
     LOG.info("Mounting {} at {}", ufsUri, alluxioPath);
 
     try (LockResource r = new LockResource(mWriteLock)) {
+      Map<String, String> properties = options.getPropertiesMap();
+      mState.applyAndJournal(journalContext, AddMountPointEntry.newBuilder()
+          .addAllProperties(properties.entrySet().stream()
+              .map(entry -> StringPairEntry.newBuilder()
+                  .setKey(entry.getKey()).setValue(entry.getValue()).build())
+              .collect(Collectors.toList()))
+          .setAlluxioPath(alluxioPath)
+          .setMountId(mountId)
+          .setReadOnly(options.getReadOnly())
+          .setShared(options.getShared())
+          .setUfsPath(ufsUri.toString())
+          .build());
+    }
+  }
+
+  /**
+   * Verify if the given (alluxioPath, ufsPath) can be inserted into MountTable.
+   * @param alluxioUri the alluxio path that is about to be the mountpoint
+   * @param ufsUri the UFS path that is about to mount
+   * @throws FileAlreadyExistsException
+   * @throws InvalidPathException
+   */
+  public void validateMountPoint(AlluxioURI alluxioUri, AlluxioURI ufsUri)
```

Comment:
I propose we make this method return a `MountTable.ValidatedPathPair` object, and make `add` accepts this object instead of arbitrary `alluxioUri` and `ufsUri` pair. This way we can avoid validating the paths multiple times. `ValidatedPathPair` will have no public constructors and can only be instanciated in `MountTable`.